File: Markup.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace party_crab {
    public class Markup
    {
        static Dictionary<string, Tuple<string, string>> defaultMarkups = new Dictionary<string, Tuple<string, string>>
        {
            { "**", new Tuple<string, string>("<b>", "</b>") },
            { "_", new Tuple<string, string>("<i>", "</i>") },
            { "*", new Tuple<string, string>("<i>", "</i>") },
            { "||", new Tuple<string, string>("<mark>", "</mark>") }
        };

        public static string MarkupReplace(string text, string value, string start, string end)
        {
            string returnText = text;
            int valueCount = returnText.Split(new string[] { value }, StringSplitOptions.None).Length - 1;
            int escapeCount = returnText.Split(new string[] { "\\" + value }, StringSplitOptions.None).Length - 1;
            valueCount -= escapeCount;

            if (valueCount % 2 != 0)
            {
                valueCount -= 1;
            }

            int valuePrevious = 0;
            for (int i = 0; i < valueCount / 2; i++)
            {
                int valueStart = returnText.IndexOf(value, valuePrevious);
                returnText = returnText.Substring(0, valueStart) + start + returnText.Substring(valueStart + value.Length);

                int valueEnd = returnText.IndexOf(value, valueStart + value.Length);
                returnText = returnText.Substring(0, valueEnd) + end + returnText.Substring(valueEnd + value.Length);

                valuePrevious = valueEnd + value.Length;
            }

            return returnText;
        }

        public static string MarkupColor(string text)
        {
            int valueCount = Math.Min(text.Split('[').Length - 1, text.Split(']').Length - 1);

            if (valueCount % 2 != 0)
            {
                valueCount -= 1;
            }

            string pattern = @"\[(.*?)\]";
            int valueAmount = Regex.Matches(text, pattern).Count;
            string returnText = Regex.Replace(text, pattern, "<color=$1>");
            returnText += string.Concat(Enumerable.Repeat("</color>", valueAmount));

            return returnText;
        }

        public static string MarkupChange(string text, Dictionary<string, Tuple<string, string>> markups = null)
        {
            string returnText = text;

            if (markups == null)
            {
                markups = defaultMarkups;
            }

            foreach (var markup in markups)
            {
                returnText = MarkupReplace(returnText, markup.Key, markup.Value.Item1, markup.Value.Item2);
            }

            returnText = MarkupColor(returnText);

            return returnText;
        }
    }
}

File: server/markup.py
import re

default_markups = {
    "**": ("<b>",    "</b>"),
    "_":  ("<i>",    "</i>"),
    "*":  ("<i>",    "</i>"),
    "||": ("<mark>", "</mark>")
}

def markup_replace(text, value, start, end):
    return_text = text
    value_count = return_text.count(value)
    escape_count = return_text.count("\\"+value)
    value_count -= escape_count

    if value_count % 2:
        value_count -= 1
    
    value_previous = 0
    for _ in range(int(value_count / 2)):
        value_start = return_text.find(value, value_previous)
        
        return_text = return_text[:value_start] + start + return_text[value_start + len(value):]
        
        value_end = return_text.find(value, value_start + len(value))
        return_text = return_text[:value_end] + end + return_text[value_end + len(value):]
        
        value_previous = value_end + len(value)
        
    return return_text


def markup_color(text):
    value_count = min(text.count("["), text.count("]"))

    if value_count % 2:
        value_count -= 1

    pattern = r"\[(.*?)\]"
    value_amount = len(re.findall(pattern, text))
    return_text = re.sub(pattern, r"<color=\g<1>>", text)
    return_text = return_text + "</color>"*value_amount

    return return_text


def markup(text, markups=default_markups):
    return_text = text
    
    for key, value in markups.items():
        return_text = markup_replace(return_text, key, value[0], value[1])
    
    return_text = markup_color(return_text)
    
    return return_text


if __name__ == "__main__":
    text1 = "normal *italic* _italic_ **bold** ||mark|| [pink]googoo gaga :D"
    print(text1)
    text1 = markup(text1)
    print(text1)
